### PR TITLE
Compute two-mode squeezing parameters

### PR DESCRIPTION
**Pull request by soranjh**

### CODE REVIEW

File: strawberryfields/gbs/vibronic.py [hunk 1]
```diff
@@ -145,27 +159,37 @@ def gbs_params(
...
-    J = Wp @ Ud @ Wi
-
-    U2, s, U1 = np.linalg.svd(J)
+    c = 299792458
+    h = 6.62607015e-34
+    k = 1.380649e-23
+
+    if T < 0:
+        raise ValueError("Temperature must be zero or positive")
+    elif T > 0:
```
Comment (soranjh) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
    if T > 0:
```

File: tests/gbs/test_vibronic.py [hunk 2]
```diff
@@ -41,7 +41,16 @@ class TestGBSParams:
     )
     d = np.array([0.2254, 0.1469, 1.5599, -0.3784, 0.4553, -0.3439, 0.0618])
 
-    U1, S, U2, alpha = vibronic.gbs_params(w, wp, Ud, d)
+    T = 300.0
+
+    t = np.array([1.19894657e-04, 6.07995705e-04, 1.25672831e-02, 3.34594410e-02, 4.15755027e-02,
+                  6.55350936e-02, 2.24605404e-01])
```
Comment (trbromley) [inline tests/gbs/test_vibronic.py]: Do we need this, since it's re-declared a few lines below?

File: tests/gbs/test_vibronic.py [hunk 3]
```diff
@@ -41,7 +41,16 @@ class TestGBSParams:
...
+
+    c = 299792458
+    h = 6.62607015e-34
+    k = 1.380649e-23
```
Comment (trbromley) [inline tests/gbs/test_vibronic.py]: Can also use `scipy.constants` here

File: tests/gbs/test_vibronic.py [hunk 4]
```diff
@@ -67,6 +76,16 @@ def test_duschinsky(self):
...
+    def test_invalid_temperature(self):
+        """Test if function raises a ``ValueError`` when a negative temperature is given."""
+        with pytest.raises(ValueError, match="Temperature must be zero or positive"):
+            vibronic.gbs_params(self.w, self.wp, self.Ud, self.d, -1)
+
+    def test_twomode(self):
+        """Test if function returns two-mode squeezing parameters that correctly reconstruct the
+        input normal mode frequencies."""
+        w = -0.02 * self.k * self.T / self.h / self.c * np.log(np.tanh(self.t))
+        assert np.allclose(w, self.w)
```
Comment (trbromley) [inline tests/gbs/test_vibronic.py]:
```suggestion
        assert np.allclose(w, self.w)

```

File: tests/gbs/test_vibronic.py [hunk 5]
```diff
@@ -67,6 +76,16 @@ def test_duschinsky(self):
         Ud = np.diag(self.wp ** -0.5) @ J @ np.diag(self.w ** 0.5)
         assert np.allclose(Ud, self.Ud)
 
+    def test_invalid_temperature(self):
+        """Test if function raises a ``ValueError`` when a negative temperature is given."""
+        with pytest.raises(ValueError, match="Temperature must be zero or positive"):
+            vibronic.gbs_params(self.w, self.wp, self.Ud, self.d, -1)
+
```
Comment (trbromley) [inline tests/gbs/test_vibronic.py]: Maybe a `test_zero_temperature` would be good, to check that the two mode squeezing parameters are zero in this case

File: tests/gbs/test_vibronic.py [hunk 6]
```diff
@@ -67,6 +76,16 @@ def test_duschinsky(self):
...
+    def test_twomode(self):
+        """Test if function returns two-mode squeezing parameters that correctly reconstruct the
+        input normal mode frequencies."""
+        w = -0.02 * self.k * self.T / self.h / self.c * np.log(np.tanh(self.t))
```
Comment (trbromley) [inline tests/gbs/test_vibronic.py]:
```suggestion
        w = - self.k * self.T / (0.5 * h * w * c * 100) * np.log(np.tanh(self.t))
```
Just to make it look more like the inverse of the function, e.g., the 0.02 confused me at first

File: tests/gbs/test_vibronic.py (hunk 5 above)
Comment (trbromley) [reply tests/gbs/test_vibronic.py]: This is also why the coverage checks are failing

image

**Comment by trbromley** [comment]
Also, this seems "Ready for review" more than "WIP/draft"

**Comment by soranjh** [comment]
> Thanks @soranjh, looks good. Left some suggestions and comments, one thing to discuss is the return order of `gbs_params`.

Thanks @trbromley for the great suggestions and corrections!

File: strawberryfields/gbs/vibronic.py [hunk 7]
```diff
@@ -109,6 +109,20 @@
...
 
+It is noted  that finite temperature FCPs can be obtained with a GBS setup explained in the
+following. First, 2N two-mode squeezed vacuum states are prepared and the first :math:`N` modes
+are treated with the same GBS operations explained for the zero K case, while the second
+:math:`N` modes are directly sent for photon number measurement. Then, the number of photons
+measured in the pairs of first and second :math:`N` modes, :math:`m_k` and :math:`n_k`,
+respectively are used to determine the transition frequencies and compute the finite temperature
+FCP. The Gaussian gate parameters for the first :math:`N` modes are obtained for a given molecule
+with the same procedure explained for the zero Kelvin case and the two-mode squeezing parameters
+(:math:`t_i`) are obtained according to the distribution of phonons in the ground electronic state
```
Comment (agran2018) [inline strawberryfields/gbs/vibronic.py]: ground electronic state => electronic ground state

File: strawberryfields/gbs/vibronic.py [hunk 8]
```diff
@@ -145,27 +159,37 @@ def gbs_params(
     >>>     ]
     >>> )
     >>> d = np.array([0.2254, 0.1469, 1.5599, -0.3784, 0.4553, -0.3439, 0.0618])
-    >>> p = gbs_params(w, wp, Ud, d)
+    >>> T = 0.0
+    >>> p = gbs_params(w, wp, Ud, d, T)
 
     Args:
         w (array): normal mode frequencies of the electronic ground state (:math:`\text{cm}^{-1}`)
         wp (array): normal mode frequencies of the electronic excited state (:math:`\text{cm}^{-1}`)
```
Comment (agran2018) [inline strawberryfields/gbs/vibronic.py]: maybe, to make the description as general as possible, you may want to refer to ``w`` and ``wp`` as the normal mode frequencies of the initial and final electronic states. Such a description would work for vibronic spectrum computed for transitions between (electronic) ground and excited states of the same molecule and between electronic ground states of a molecule wit different charges.

File: strawberryfields/gbs/vibronic.py [hunk 9]
```diff
@@ -145,27 +159,37 @@ def gbs_params(
     >>>     ]
     >>> )
     >>> d = np.array([0.2254, 0.1469, 1.5599, -0.3784, 0.4553, -0.3439, 0.0618])
-    >>> p = gbs_params(w, wp, Ud, d)
+    >>> T = 0.0
+    >>> p = gbs_params(w, wp, Ud, d, T)
 
     Args:
         w (array): normal mode frequencies of the electronic ground state (:math:`\text{cm}^{-1}`)
```
Comment (trbromley) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
        w (array): normal mode frequencies of the electronic ground state (:math:`\mbox{cm}^{-1}`)
```
I realised this doesn't render correctly

File: strawberryfields/gbs/vibronic.py (hunk 8 above)
Comment (trbromley) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
        wp (array): normal mode frequencies of the electronic excited state (:math:`\mbox{cm}^{-1}`)
```

File: strawberryfields/gbs/vibronic.py (hunk 7 above)
Comment (soranjh) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
(:math:`t_i`) are obtained according to the distribution of phonons in the electronic ground state
```

File: strawberryfields/gbs/vibronic.py [hunk 10]
```diff
@@ -145,27 +159,37 @@ def gbs_params(
...
         wp (array): normal mode frequencies of the electronic excited state (:math:`\text{cm}^{-1}`)
         Ud (array): Duschinsky matrix
         d (array): Duschinsky displacement vector corrected with wp
+        T (float): temperature
```
Comment (soranjh) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
        T (float): temperature (Kelvin)
```

File: strawberryfields/gbs/vibronic.py [hunk 11]
```diff
@@ -145,27 +159,37 @@ def gbs_params(
...
 
     Returns:
-        tuple[array, array, array, array]: the first interferometer unitary matrix :math:`U_{1}`,
-        the squeezing parameters :math:`r`, the second interferometer unitary matrix :math:`U_{2}`,
-        and finally the displacement parameters :math:`\alpha`
+        tuple[array, array, array, array, array]: the first interferometer unitary matrix
+        :math:`U_{1}`, the squeezing parameters :math:`r`, the second interferometer unitary
+        matrix :math:`U_{2}`, the displacement parameters :math:`\alpha`, and finally the
+        two-mode squeezing parameters :math:`t`
```
[the remark this answers is not in the document]
Comment (soranjh) [reply strawberryfields/gbs/vibronic.py]: I will change the order. I think ``sample.vibronic`` does not need to have T.

File: strawberryfields/gbs/vibronic.py [hunk 12]
```diff
@@ -145,27 +159,37 @@ def gbs_params(
...
     """
-    Wi = np.diag(w ** -0.5)
-    Wp = np.diag(wp ** 0.5)
-    J = Wp @ Ud @ Wi
-
-    U2, s, U1 = np.linalg.svd(J)
+    c = 299792458
+    h = 6.62607015e-34
+    k = 1.380649e-23
```
[the remark this answers is not in the document]
Comment (soranjh) [reply strawberryfields/gbs/vibronic.py]: I will get them from scipy. Prefer to use the SI unit for h.

File: strawberryfields/gbs/vibronic.py [hunk 13]
```diff
@@ -145,27 +159,37 @@ def gbs_params(
...
+
+    if T < 0:
+        raise ValueError("Temperature must be zero or positive")
+    if T > 0:
+        t = np.arctanh(np.exp(-0.5 * h * (w * c * 100) / k / T))
```
[the remark this answers is not in the document]
Comment (soranjh) [reply strawberryfields/gbs/vibronic.py]: It seems that I have used \omega for both wavenumber and angular frequency. Will correct that. c is definitely needed here to convert the wavenumber to frequency.

File: tests/gbs/test_vibronic.py (hunk 2 above)
Comment (soranjh) [reply tests/gbs/test_vibronic.py]: Will remove it.

File: tests/gbs/test_vibronic.py (hunk 3 above)
Comment (soranjh) [reply tests/gbs/test_vibronic.py]: yes.

File: strawberryfields/gbs/vibronic.py (hunk 7 above)
Comment (soranjh) [reply strawberryfields/gbs/vibronic.py]: done.

File: strawberryfields/gbs/vibronic.py (hunk 8 above)
Comment (soranjh) [reply strawberryfields/gbs/vibronic.py]: done.

File: strawberryfields/gbs/vibronic.py [hunk 14]
```diff
@@ -109,6 +109,22 @@
...
 
+It is noted that finite temperature FCPs can be obtained with a GBS setup explained in the
+following. First, :math:`2N` two-mode squeezed vacuum states are prepared and the first :math:`N` modes
+are treated with the same GBS operations explained for the zero Kelvin case, while the second
+:math:`N` modes are directly sent for photon number measurement. Then, the number of photons
+measured in the pairs of first and second :math:`N` modes, :math:`m_k` and :math:`n_k`,
+respectively are used to determine the transition frequencies and compute the finite temperature
+FCP. The Gaussian gate parameters for the first :math:`N` modes are obtained for a given molecule
+with the same procedure explained for the zero Kelvin case and the two-mode squeezing parameters
+(:math:`t_i`) are obtained according to the distribution of phonons in the electronic ground state
```
Comment (ixfoduap) [inline strawberryfields/gbs/vibronic.py]: change to: "in the vibrational levels of the electronic ground state.."

File: strawberryfields/gbs/vibronic.py [hunk 15]
```diff
@@ -148,24 +166,29 @@ def gbs_params(
...
+        parameters :math:`\alpha`
     """
-    Wi = np.diag(w ** -0.5)
-    Wp = np.diag(wp ** 0.5)
-    J = Wp @ Ud @ Wi
-
-    U2, s, U1 = np.linalg.svd(J)
+    if T < 0:
+        raise ValueError("Temperature must be zero or positive")
+    if T > 0:
```
Comment (ixfoduap) [inline strawberryfields/gbs/vibronic.py]: I wonder if we get any errors if a user inputs an extremely low temperature, for example `T=1e-20` ?

File: tests/gbs/test_vibronic.py [hunk 16]
```diff
@@ -67,6 +71,22 @@ def test_duschinsky(self):
...
+    def test_twomode(self):
+        """Test if function returns two-mode squeezing parameters that correctly reconstruct the
+        input normal mode frequencies."""
+        w = -k * self.T / (0.5 * h * c * 100) * np.log(np.tanh(self.t))
+        assert np.allclose(w, self.w)
+
+    def test_zero_temperature(self):
+        """Test if function returns zero two-mode squeezing parameters when temperature is zero."""
+        t, _, _, _, _ = vibronic.gbs_params(self.w, self.wp, self.Ud, self.d, 0.0)
+        assert np.all(t == 0)
```
Comment (ixfoduap) [inline tests/gbs/test_vibronic.py]: why do we need `np.all()` here?

Comment (trbromley) [reply tests/gbs/test_vibronic.py]: `t` is an N dimensional numpy array so `t==0` evaluates to something like `[True, True, False, True, True]`
The all acts on a list of bools and only returns true if all the items in the list are true

File: strawberryfields/gbs/vibronic.py (hunk 15 above)
Comment (soranjh) [reply strawberryfields/gbs/vibronic.py]: No.

File: tests/gbs/test_vibronic.py (hunk 16 above)
Comment (soranjh) [reply tests/gbs/test_vibronic.py]: because t is an array.

File: strawberryfields/gbs/vibronic.py (hunk 14 above)
Comment (soranjh) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
(:math:`t_i`) are obtained according to the distribution of phonons in the vibrational levels of the
```

File: strawberryfields/gbs/vibronic.py [hunk 17]
```diff
@@ -109,6 +109,22 @@
...
+It is noted that finite temperature FCPs can be obtained with a GBS setup explained in the
+following. First, :math:`2N` two-mode squeezed vacuum states are prepared and the first :math:`N` modes
+are treated with the same GBS operations explained for the zero Kelvin case, while the second
+:math:`N` modes are directly sent for photon number measurement. Then, the number of photons
+measured in the pairs of first and second :math:`N` modes, :math:`m_k` and :math:`n_k`,
+respectively are used to determine the transition frequencies and compute the finite temperature
+FCP. The Gaussian gate parameters for the first :math:`N` modes are obtained for a given molecule
+with the same procedure explained for the zero Kelvin case and the two-mode squeezing parameters
+(:math:`t_i`) are obtained according to the distribution of phonons in the electronic ground state
+vibrational levels according to a Boltzmann distribution with Boltzmann factor:
```
Comment (soranjh) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
electronic ground state according to a Boltzmann distribution with Boltzmann factor:
```

File: tests/gbs/test_vibronic.py (hunk 16 above)
Comment (ixfoduap) [reply tests/gbs/test_vibronic.py]: My bad, thought `t` was a float

File: strawberryfields/gbs/vibronic.py [hunk 18]
```diff
@@ -109,6 +109,22 @@
 geometries, vibrational frequencies and normal coordinates of the molecule in its (electronic)
 ground and excited states.
 
+Finite temperature FCPs can also be obtained using a GBS setup.
+following. First, :math:`N` two-mode squeezed vacuum states are prepared and the first :math:`N` modes
```
Comment (ixfoduap) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
First, :math:`N` two-mode squeezed vacuum states are prepared and the first :math:`N` modes
```

File: strawberryfields/gbs/vibronic.py [hunk 19]
```diff
@@ -109,6 +109,22 @@
...
+are treated with the same GBS operations explained for the zero Kelvin case. The second
+:math:`N` modes are directly sent for photon number measurement. The number of photons
+measured in the first and second modes, :math:`m_k` and :math:`n_k`,
+respectively, are used to determine the transition frequencies and compute the finite temperature
+FCP. The Gaussian gate parameters for the first :math:`N` modes are obtained for a given molecule
+with the same procedure explained for the zero Kelvin case, and the two-mode squeezing parameters
```
Comment (ixfoduap) [inline strawberryfields/gbs/vibronic.py]:
```suggestion
with the same procedure explained for the zero Kelvin case. The two-mode squeezing parameters
```

**Pull request merged**